### PR TITLE
feat: speedup initializing MetaMask

### LINKED ISSUE #120: How avoid import seed to metamask every time
When I bootstrap dappeteer,it need import seed,add network,switch network,it seems too slow.Futhermore,it failed sometimes when add new network to metamask.Are there any way to launch by my own prepared metamask to avoid these duplicate steps?

### PR DESCRIPTION
**Pull request by BeroBurny**
**Short description of work done**
Implement methods to backup user profile data and load the browser from updated data to speed up the initial start time of dappateer

### PR Checklist

- [x] I have run linter locally
- [x] I have run unit and integration tests locally
  - [ ] Update configuration the newest version (readme and const)
- [x] Rebased to master branch / merged master

Closes #120

### CODE REVIEW

**Comment by BeroBurny** [comment]
Currently is draft as i need to write documentation for a new feature but code can be check before it 

cc @irubido @Lykhoyda @mpetrunic

**Comment by Lykhoyda** [comment]
I have to ask a nerdy question, what is the time difference between using the implemented method and without it ? 🤓

**Comment by BeroBurny** [comment]
> I have to ask a nerdy question, what is the time difference between using the implemented method and without it ? nerd_face

result from test
```ssh
  userData
    MetaMask
      ✔ should successfully store user data (20185ms) note: there is 1sec pause
      ✔ should successfully launch from custom user folder (8899ms)
    Flask
      ✔ should successfully store user data (21544ms) note: there is 1sec pause
      ✔ should successfully launch from custom user folder (9719ms)
    Predefined
      ✔ should successfully launch project's default MetaMask (8959ms)
      ✔ should successfully launch project's default Flask (10046ms)

```

File: src/setup/utils/addKeyToMetaMaskManifest.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import * as path from "path";
+import { readJsonSync, writeJsonSync } from "fs-extra";
```
[the remark this answers is not in the document]
Comment (BeroBurny) [reply src/setup/utils/addKeyToMetaMaskManifest.ts:2]: `fs-extra` is already part of the dependency (used for copy files from user to destination and vice versa)

File: src/setup/launch.ts [hunk 2]
```diff
@@ -54,25 +58,35 @@ export async function launch(
     metamaskPath = options.metaMaskPath;
   }
 
+  addKeyToMetaMaskManifest(metamaskPath, options.key);
```
[the remark this answers is not in the document]
Comment (BeroBurny) [reply src/setup/launch.ts:61]: It adds `key` field to `manifest.json` file of downloaded extension from GitHub.

> Why do we need a key?

To keep the same extension hash id between browsers and operative systems we need key

> When you load an extension in "unpacked mode", an ID is automatically generated in an unpredictable way. The only way to control the extension ID during development is by setting the ["key"](https://developer.chrome.com/apps/manifest/key) field in the manifest file, as the [documentation suggests](https://developer.chrome.com/apps/app_identity#copy_key).

source [https://stackoverflow.com/a/21500707](https://stackoverflow.com/a/21500707)

File: src/setup/utils/addKeyToMetaMaskManifest.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+const DEFAULT_KEY =
+  "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAjamYPwVQXybbRpzu+eIfPAy6WGb0dLqIEemi9CKTshabZg03Nqe7kqaMEWkHMWAk5MFgIhApMV7axEoidyWAlx/Vp5F9lNczfF8Nb1FaujNZKNnhkq98y2O6/PKK5oeHZAc/qwx3JFRBPRGdSE/cEeQzNGXRycnZucR3VAVMt3NI7DoGevCj+CYbqTITdVr7SqmiG8wfeIvIBOPJXo5jrNhjZ5de31mRYG8utDDd5cZN9dAG2ijKDnTTrvUBT3DomiuzzRhWAtSldvpQ0EJ/5rNEIkoStC1o5jKu5E7LQWqe+92My7YoaRPdIOjHmStlw6HKzjBfIloKFjv+9QIKqQIDAQAB";
```
[the remark this answers is not in the document]
Comment (BeroBurny) [reply src/setup/utils/addKeyToMetaMaskManifest.ts:5]: Key from the google extension store (it is from MetaMask), reason explained in comment for `addKeyToMetaMaskManifest`

File: src/constants.ts [hunk 4]
```diff
@@ -1 +1,15 @@
+import path from "path";
+import { getDappateerPath } from "./helpers/utils";
+
 export const EXAMPLE_WEBSITE = "http://example.org";
+
+export const RECOMMENDED_METAMASK_VERSION = "v10.23.0";
+
+export const DEFAULT_METAMASK_USERDATA = path.join(
+  getDappateerPath(),
+  "userData/chrome-mm"
```
[the remark this answers is not in the document]
Comment (BeroBurny) [reply src/constants.ts:10]: there is no need, and default user data will follow the version from dappeteer.

File: src/setup/setupMetaMask.ts [hunk 5]
```diff
@@ -59,6 +60,27 @@ export async function setupMetaMask<Options = MetaMaskOptions>(
...
+  browser: DappeteerBrowser,
+  password: string
+): Promise<Dappeteer> {
+  const page = await getMetaMaskPage(browser);
+  const metaMask = await getMetaMask(page);
+
+  await metaMask.page.evaluate(() => {
+    (window as unknown as { signedIn: boolean }).signedIn = false;
+  });
+  await page.waitForTimeout(100);
```
[the remark this answers is not in the document]
Comment (BeroBurny) [reply src/setup/setupMetaMask.ts:73]: Some internal parts need that state, by default is true but when you bootstrap MetaMask from a state extension will be locked

[the remark this answers is not in the document]
Comment (mpetrunic) [reply src/setup/setupMetaMask.ts:73]: I don't understand how your comment relates to timeout haha

[the remark this answers is not in the document]
Comment (BeroBurny) [reply src/setup/setupMetaMask.ts:73]: Ah GH, this that you are talking about `evaluate`, for some reason this timeout is required for a test not to be flaky.
For some reason, without it, Dappateer can miss reading a state

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -33,6 +33,8 @@ jobs:
         with:
           name: debug_screenshots
           path: ./*.png
+      - name: Tests UserData
+        run: 'xvfb-run --auto-servernum --server-args="-screen 0, 1920x1080x24" yarn run test:${{matrix.automation}}:userData --timeout 240000'
```
Comment (mpetrunic) [inline .github/workflows/main.yml:37]: We no longer need xvfb

Comment (BeroBurny) [reply .github/workflows/main.yml:37]: forget to update CI :sweat:

**Review by irubido — APPROVED** [review]

**Review by irubido — DISMISSED** [review]

File: src/browser.ts [hunk 7]
```diff
@@ -14,4 +14,8 @@ export interface DappeteerBrowser<Browser = unknown, Page = unknown>
   close(): Promise<void>;
 
   wsEndpoint(): string;
+
+  getUserDataDirPath(): string;
+
+  storeUserData(destination: string): boolean;
```
Comment (Lykhoyda) [inline src/browser.ts:20]: How are we using the boolean returned from the method?

File: test/flask/snaps.spec.ts [hunk 8]
```diff
@@ -8,6 +8,10 @@ describe("snaps", function () {
   let metaMaskPage: DappeteerPage;
 
   before(function (this: TestContext) {
+    if (Boolean(process.env.USER_DATA_TEST) === true) {
```
Comment (Lykhoyda) [inline test/flask/snaps.spec.ts:11]: probably we can move the check to the helper method as the same code is used multiple times

File: src/browser.ts (hunk 7 above)
Comment (BeroBurny) [reply src/browser.ts:20]: It returns `true` if success `false` if fails somewhere
For future proofing and for users

**Review by Lykhoyda — APPROVED** [review]

**Pull request merged**